apps/settings/js/modules/bluetooth/bluetooth_connection_manager.js

```diff
@@ -0,0 +1,868 @@
+/**
+ * BluetoothConnectionManager:
+ * BluetoothConnectionManager only update state and does not involve in any UI
+ *   logic.
+ *
+ * @module BluetoothConnectionManager
+ */
+define(function(require) {
+  'use strict';
+
+  var AdapterManager = require('modules/bluetooth/bluetooth_adapter_manager');
+  var AsyncStorage = require('shared/async_storage');
+  require('shared/async_storage');
+
+  var _debug = false;
+  var Debug = function() {};
+  if (_debug) {
+    Debug = function btam_debug(msg) {
+      console.log('--> [BluetoothConnectionManager]: ' + msg);
+    };
+  }
+
+  /**
+   * @alias module:modules/bluetooth/BluetoothConnectionManager
+   * @requires module:modules/bluetooth/bluetooth_adapter_manager
+   * @requires module:shared/async_storage
+   * @return {BluetoothConnectionManager}
+   */
+  var BluetoothConnectionManager = {
+    /**
+     * The profiles of connected device that we are defined here.
+     *
+     * @public
+     * @memberOf BluetoothConnectionManager
+     * @type {Object}
+     */
+    Profiles: {
+      'hfp': 0x111E,
+      'a2dp': 0x110D
+    },
+
+    /**
+     * The address of device that we are trying to connect.
+     *
+     * @public
+     * @memberOf BluetoothConnectionManager
+     * @type {String}
+     */
+    connectingAddress: null,
+
+    /**
+     * A object that we cache the connected devices information(address, device,
+     * connectedProfiles). It will be inited while default adapter is ready.
+     * And these information is coming from profile events.
+     * Each connected device is hashed by device address.
+     *
+     * EX:
+     * _connectedDevicesInfo = {
+     *   'AA:BB:CC:00:11:22': {
+     *     'device': DeviceObject,
+     *     'connectedProfiles': {
+     *       'hfp': true,
+     *       'a2dp': false
+     *     }
+     *   }
+     * };
+     *
+     * @private
+     * @memberOf BluetoothConnectionManager
+     * @type {Object}
+     */
+    _connectedDevicesInfo: {},
+
+    /**
+     * Return the cache of connected devices in ConnectionManager.
+     *
+     * @access public
+     * @memberOf BluetoothConnectionManager
+     * @return {Object}
+     */
+    getConnectedDevices: function btcm_getConnectedDevices() {
+      return this._connectedDevicesInfo;
+    },
+
+    /**
+     * These listeners would be called when device is connecting.
+     *
+     * @memberOf BluetoothConnectionManager
+     * @type {Array}
+     */
+    _connectingListeners: [],
```

Comment:
This could be implemented is a easier and more flexible way. We can have a object like:
```
_listeners: {
  'connecting': [],
  'connected': [],
  'disconnected': [],
  'profileChanged': []
}
```

With this object, `addEventListener`, `removeEventListener`, and the function for emitting the events can be implemented like:
```
addEventListener: function(eventName, callback) {
  // TODO: check eventName is valid
  // TODO: check callback is valid
  this._listeners[eventName].push(callback);
}
```